[zh-cn]: Update translation of Promise (#13809)
mdn/translated-content: files/zh-cn/web/javascript/reference/global_objects/promise/index.md

{{JSRef}}

**`Promise`** 对象表示异步操作最终的完成（或失败）以及其结果值。

若想了解 promise 的工作方式以及如何使用它们，我们建议你先阅读[使用 Promise](/zh-CN/docs/Web/JavaScript/Guide/Using_promises)。

## 描述

一个 **`Promise`** 是一个代理，它代表一个在创建 promise 时不一定已知的值。它允许你将处理程序与异步操作的最终成功值或失败原因关联起来。这使得异步方法可以像同步方法一样返回值：异步方法不会立即返回最终值，而是返回一个 _promise_，以便在将来的某个时间点提供该值。

一个 `Promise` 必然处于以下几种状态之一：

- *待定（pending）*：初始状态，既没有被兑现，也没有被拒绝。
- *已兑现（fulfilled）*：意味着操作成功完成。
- *已拒绝（rejected）*：意味着操作失败。

> 一个待定的 Promise *最终状态*可以是*已兑现*并返回一个 value，或者是*已拒绝*并返回一个 reason（错误）。当其中任意一种情况发生时，通过 Promise 的 `then` 方法串联的处理程序将被调用。如果绑定相应处理程序时 Promise 已经兑现或拒绝，这处理程序将被立即调用，因此在异步操作完成和绑定处理程序之间不存在竞态条件。

Reviewer: why we don't translate 'value' and 'reason'?

Author: In the Promise.allSettled method returns an array of promise results - fulfilled return `{status: 'fulfilled', value: the value}` and rejected return `{status: 'rejected', reason: the reason}`. So I am considering not translating the `value` and `reason` because they have means of variables.

Reviewer: Yes, the "reason" here would be _set_ as the `reason` key in the return object of `allSettled` (if we call the method). But this is a general description. We haven't given any object or function signature here. And the `reason` key should mean "reason", that is _why_ the promise is rejected. I don't think only using English here would bring more meaning here.

Author: Agree with that.

> 一个待定的 Promise *最终状态*可以是*已兑现*并返回一个 value，或者是*已拒绝*并返回一个 reason（错误）。当其中任意一种情况发生时，通过 Promise 的 `then` 方法串联的处理程序将被调用。如果绑定相应处理程序时 Promise 已经兑现或拒绝，这处理程序将被立即调用，因此在异步操作完成和绑定处理程序之间不存在竞态条件。

Author: ```suggestion
一个待定的 Promise *最终状态*可以是*已兑现*并返回一个值，或者是*已拒绝*并返回一个原因（错误）。当其中任意一种情况发生时，通过 Promise 的 `then` 方法串联的处理程序将被调用。如果绑定相应处理程序时 Promise 已经兑现或拒绝，这处理程序将被立即调用，因此在异步操作完成和绑定处理程序之间不存在竞态条件。
```

如果一个 Promise 已经被兑现或拒绝，即不再处于待定状态，那么则称之为已*敲定（settled）*。

![流程图展示了 Promise 状态在 `pending`、`fulfilled` 和 `rejected` 之间如何通过 `then()` 和 `catch()` 处理程序进行转换。一个待定的 Promise 可以变成已兑现或已拒绝的状态。如果 Promise 已经兑现，则会执行“on fulfillment”处理程序（即 `then()` 方法的第一个参数），并继续执行进一步的异步操作。如果 Promise 被拒绝，则会执行错误处理程序，可以将其作为 `then()` 方法的第二个参数或 `catch()` 方法的唯一参数来传递。](promises.png)

> 你还会听到使用 _resolved_ 这个术语来描述 Promise——这意味着该 Promise 已经敲定（settled），或为了匹配另一个 Promise 的最终状态而被“锁定（lock-in）”，进一步解决或拒绝它都没有影响。原始 Promise 提案中的 [States and fates](https://github.com/domenic/promises-unwrapping/blob/master/docs/states-and-fates.md) 文档包含了更多关于 Promise 术语的细节。在口语中，“resolved”的 Promise 通常等价于“fulfilled”的 Promise，但是正如“States and fates”所示，已解决的 Promise 也可以是待定或拒绝的。例如：

Reviewer: I think the term "_resolved_" is the same in English document:

> You will also hear the term resolved used with promises — this means that the promise is settled or "locked-in" to match the eventual state of another promise, and further resolving or rejecting it has no effect. The States and fates document from the original Promise proposal contains more details about promise terminology. Colloquially, "resolved" promises are often equivalent to "fulfilled" promises, but as illustrated in "States and fates", resolved promises can be pending or rejected as well. For example:

But why using English word and then using "已解决".

> 你还会听到使用 _resolved_ 这个术语来描述 Promise——这意味着该 Promise 已经敲定（settled），或为了匹配另一个 Promise 的最终状态而被“锁定（lock-in）”，进一步解决或拒绝它都没有影响。原始 Promise 提案中的 [States and fates](https://github.com/domenic/promises-unwrapping/blob/master/docs/states-and-fates.md) 文档包含了更多关于 Promise 术语的细节。在口语中，“resolved”的 Promise 通常等价于“fulfilled”的 Promise，但是正如“States and fates”所示，已解决的 Promise 也可以是待定或拒绝的。例如：

Author: ```suggestion
你还会听到使用*已解决（resolved）*这个术语来描述 Promise——这意味着该 Promise 已经敲定（settled），或为了匹配另一个 Promise 的最终状态而被“锁定（lock-in）”，进一步解决或拒绝它都没有影响。原始 Promise 提案中的 [States and fates](https://github.com/domenic/promises-unwrapping/blob/master/docs/states-and-fates.md) 文档包含了更多关于 Promise 术语的细节。在口语中，“已解决”的 Promise 通常等价于“已兑现”的 Promise，但是正如“States and fates”所示，已解决的 Promise 也可以是待定或拒绝的。例如：
```

```js
new Promise((resolveOuter) => {
  resolveOuter(
    new Promise((resolveInner) => {
      setTimeout(resolveInner, 1000);
    }),
  );
});
```

> 此 Promise 在创建时已经被解决（因为 `resolveOuter` 是同步调用的），但它被解决为另一个 Promise，因此在内部 Promise 实现后的 1 秒钟之后才会被*兑现*。在实践中，“解决”通常是在幕后完成的，不可观察，只有其兑现或拒绝是可观察的。

Reviewer: ```suggestion
此 Promise 在创建时已经被解决（因为 `resolveOuter` 是同步调用的），但它是用另一个 Promise 解决的，因此在内部 Promise 兑现的 1 秒之后才会*被兑现*。在实践中，“解决”过程通常是在幕后完成的，不可观察，只有其兑现或拒绝是可观察的。
```

> 此 Promise 在创建时已经被解决（因为 `resolveOuter` 是同步调用的），但它被解决为另一个 Promise，因此在内部 Promise 实现后的 1 秒钟之后才会被 _fulfilled_。在实践中，“解决”通常是在幕后完成的，不可观察，只有其兑现或拒绝是可观察的。

Reviewer: ```suggestion
此 Promise 在创建时已经被解决（因为 `resolveOuter` 是同步调用的），但它被解决为另一个 Promise，因此在内部 Promise 实现后的 1 秒钟之后才会被*兑现*。在实践中，“解决”通常是在幕后完成的，不可观察，只有其兑现或拒绝是可观察的。
```

> > **备注：** 其他几种语言也有一些机制来实现惰性求值和延迟计算，它们也称之为“promise”，例如Scheme。在JavaScript中，Promise 代表已经在进行中的进程，而且可以通过回调函数实现链式调用。如果你想要实现惰性求值，考虑使用不带参数的函数，例如 `f = () => expression` 来创建惰性求值表达式，然后使用 `f()` 立即求值。

Reviewer: ```suggestion
> **备注：** 其他几种语言也有一些机制来实现惰性求值和延迟计算，它们也称之为“promise”，例如 Scheme。在 JavaScript 中，Promise 代表已经在进行中的进程，而且可以通过回调函数实现链式调用。如果你想要实现惰性求值，考虑使用不带参数的函数，例如 `f = () => expression` 来创建惰性求值表达式，然后使用 `f()` 立即求值。
```

### Promise 的链式调用

> {{jsxref("Promise.prototype.then()")}}、{{jsxref("Promise.prototype.catch()")}} 和 {{jsxref("Promise.prototype.finally()")}} 方法用于将进一步的操作与已敲定的 Promise 相关联。由于这些方法返回 Promise，因此它们可以进行链接操作。

Reviewer: ```suggestion
{{jsxref("Promise.prototype.then()")}}、{{jsxref("Promise.prototype.catch()")}} 和 {{jsxref("Promise.prototype.finally()")}} 方法用于将进一步的操作与已敲定的 Promise 相关联。由于这些方法返回 Promise，因此它们可以被链式调用。
```

`.then()` 方法最多接受两个参数；第一个参数是 Promise 兑现时的回调函数，第二个参数是 Promise 拒绝时的回调函数。每个 `.then()` 返回一个新生成的 Promise 对象，这个对象可被用于链式调用，例如：

```js
const myPromise = new Promise((resolve, reject) => {
  setTimeout(() => {
    resolve("foo");
  }, 300);
});

myPromise
  .then(handleFulfilledA, handleRejectedA)
  .then(handleFulfilledB, handleRejectedB)
  .then(handleFulfilledC, handleRejectedC);
```

> 即使 `.then()` 缺少返回 Promise 对象的回调函数，处理程序仍会继续到链的下一个链式调用。因此，在最终的 `.catch()` 之前，可以安全地省略每个链式调用中处理已拒绝状态的回调函数。

Reviewer: ```suggestion
即使 `.then()` 缺少返回 Promise 对象的回调函数，处理程序仍会继续到链的下一个链式调用。因此，在最终的 `.catch()` 之前，可以安全地省略每个链式调用中处理被拒绝状态的回调函数。
```

Author: Yes, It is more appropriate to use "被" in some contexts, but if the state of `Promise` need to be emphasized, '已拒绝（的）' or '拒绝（的）' may be more proper.

> 在每个 `.then()` 中处理拒绝的 Promise 对于 Promise 链下游有重要的影响。有时候别无选择，因为有的错误必须立即被处理。在这种情况下，必须抛出某种类型的错误以维护链中的错误状态。另一方面，在没有迫切需要的情况下，最好将错误处理留到最后一个 `.catch()` 语句。`.catch()` 其实就是一个没有为 Promise 时的回调函数留出空位的 `.then()`。

Reviewer: adding "被" (may be better than "已")

```suggestion
在每个 `.then()` 中处理被拒绝的 Promise 对于 Promise 链的下游有重要的影响。有时候别无选择，因为有的错误必须立即被处理。在这种情况下，必须抛出某种类型的错误以维护链中的错误状态。另一方面，在没有迫切需要的情况下，最好将错误处理留到最后一个 `.catch()` 语句。`.catch()` 其实就是一个没有为 Promise 时的回调函数留出空位的 `.then()`。
```

```js
myPromise
  .then(handleFulfilledA)
  .then(handleFulfilledB)
  .then(handleFulfilledC)
  .catch(handleRejectedAny);
```

使用{{JSxRef("Functions/Arrow_functions", "箭头函数", "", 1)}}作为回调函数，实现 Promise 的链式调用的示例如下：

```js
myPromise
  .then((value) => `${value} and bar`)
  .then((value) => `${value} and bar again`)
  .then((value) => `${value} and again`)
  .then((value) => `${value} and again`)
  .then((value) => {
    console.log(value);
  })
  .catch((err) => {
    console.error(err);
  });
```

> **备注：** 为了更快的执行，最好将所有同步操作都放在一个处理程序中，否则如果将它们拆分为多个处理程序，执行所有处理程序将需要几个时钟周期。

> 一个 Promise 的终止条件决定了链中下一个 Promise 的“settled”状态。“fulfilled”状态表示 Promise 成功完成，而“rejected”状态表示 Promise 执行失败。链中每个已兑现的 Promise 的返回值会传递给下一个 `.then()`，而已拒绝的 Promise 会把失败原因传递给链中下一个拒绝处理函数。

Reviewer: Why not translate those terms

> 一个 Promise 的终止条件决定了链中下一个 Promise 的“settled”状态。“fulfilled”状态表示 Promise 成功完成，而“rejected”状态表示 Promise 执行失败。链中每个已兑现的 Promise 的返回值会传递给下一个 `.then()`，而已拒绝的 Promise 会把失败原因传递给链中下一个拒绝处理函数。

Author: ```suggestion
一个 Promise 的终止条件决定了链中下一个 Promise 的“已敲定”状态。“已兑现”状态表示 Promise 成功完成，而“已拒绝”状态表示 Promise 执行失败。链中每个已兑现的 Promise 的返回值会传递给下一个 `.then()`，而已拒绝的 Promise 会把失败原因传递给链中下一个拒绝处理函数。
```

- {{jsxref("Promise.any()")}}
  - : 接受一个 Promise 可迭代对象作为输入，并返回单个 `Promise`。返回的 Promise 在任何输入的 Promise 兑现时兑现，其值为第一个兑现的值。如果所有输入的 Promise 都被拒绝（包括传入的可迭代对象为空时），返回的 Promise 将被拒绝，并包含一个包含拒绝原因的 {{jsxref("AggregateError")}} 数组。
- {{jsxref("Promise.race()")}}
  - : 接受一个 Promise 可迭代对象作为输入，并返回单个 `Promise`。返回的 Promise 与第一个敲定的 Promise 的最终状态保持一致。
- {{jsxref("Promise.reject()")}}
  - : 返回一个新的 `Promise` 对象，该对象以给定的原因拒绝。
- {{jsxref("Promise.resolve()")}}

  - : 返回一个新的 `Promise` 对象，该对象以给定的值而兑现。如果值是一个 thenable 对象（即具有`then`方法），则返回的 Promise 对象会“跟随”该 thenable 对象，采用其最终的状态；否则，返回的 Promise 对象会以该值兑现。

    通常，如果您不知道一个值是否是 Promise，那么最好使用 {{jsxref("Promise.resolve", "Promise.resolve(value)")}} 将其转换成Promise 对象，并将返回值作为 Promise 来处理。

## 实例属性

这些属性定义在 `Promise.prototype` 上，由所有的 `Promise` 实例共享。

- {{jsxref("Object/constructor", "Promise.prototype.constructor")}}
  - : 创建实例对象的构造函数。对于 `Promise` 实例，初始值为 {{jsxref("Promise/Promise", "Promise")}} 构造函数。
- `Promise.prototype[@@toStringTag]`
  - : [`@@toStringTag`](/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/Symbol/toStringTag) 属性的初始值为字符串 `"Promise"`。该属性用于 {{jsxref("Object.prototype.toString()")}}。

## 实例方法

- {{jsxref("Promise.prototype.catch()")}}

>   - : 将一个拒绝处理回调函数附加到 Promise 上，并返回一个新的 Promise，如果回调被调用，则解析为回调的返回值，如果 Promise 被兑现，解析为其原始完成值。

Reviewer: ```suggestion
  - : 将一个拒绝处理回调函数附加到 Promise 上，并返回一个新的 Promise，如果回调被调用，则解决为回调的返回值，如果 Promise 被兑现，解决为其原始兑现值。
```

- {{jsxref("Promise.prototype.finally()")}}

>   - : 将一个处理程序附加到 Promise 上，并返回一个新的 Promise，当原始 Promise 被解决时解析。无论 Promise 是否被兑现，处理程序都会在 Promise 解决时被调用。

Reviewer: ```suggestion
  - : 将一个处理器附加到 Promise 上，并返回一个新的 Promise，当原始 Promise 被解决时解决。无论 Promise 是否被兑现还是被拒绝，处理器都会在 Promise 敲定时被调用。
```

- {{jsxref("Promise.prototype.then()")}}

>   - : 将一个完成处理和拒绝处理附加到 Promise 上，并返回一个新的 Promise，解析为调用处理程序的返回值，或者如果 Promise 没有被处理（即相关处理程序 `onFulfilled` 或 `onRejected` 不是函数），则解析为其原始解决值。

Reviewer: ```suggestion
  - : 将一个兑现处理器和拒绝处理器附加到 Promise 上，并返回一个新的 Promise，解决为调用处理器得到的返回值，或者如果 Promise 没有被处理（即相关处理器 `onFulfilled` 或 `onRejected` 不是函数），则以原始敲定值解决。
```

## 示例

### 基础示例

```js
const myFirstPromise = new Promise((resolve, reject) => {
  // 当异步操作成功时，我们调用 resolve(...)，当其失败时，调用 reject(...)。
  // 在这个例子中，我们使用 setTimeout(...) 来模拟异步代码。

>   // 在实际情况中，你可能会使用类似 XHR 或 HTML API 等其他工具。

Reviewer: ```suggestion
  // 在实际情况中，你可能会使用类似 XHR 或 HTML API 等。
```

  setTimeout(() => {

>     resolve("Success!"); // Yay! Everything went well!

Reviewer: ```suggestion
    resolve("成功！"); // 耶！一切顺利！
```

  }, 250);
});

myFirstPromise.then((successMessage) => {
  // successMessage 是我们在上面的 resolve(...) 函数中传入的任何内容。

>   // 它不一定是一个字符串，但如果它只是一个成功的消息，那么它可能是一个字符串。

Reviewer: ```suggestion
  // 它不一定是字符串，但如果它只是一个成功的消息，那么它大概率是字符串。
```

>   console.log(`Yay! ${successMessage}`);

Reviewer: ```suggestion
  console.log(`耶！${successMessage}`);
```

```js
const promiseA = new Promise((resolve, reject) => {
  resolve(777);
});
// 此时，“promiseA” 已经敲定了
promiseA.then((val) => console.log("异步日志记录有值：", val));
console.log("立即记录");

// 按以下顺序产生输出：
// 立即记录
// 异步日志记录有值：777
```

### Thenable

在 Promise 成为 JavaScript 语言的一部分之前，JavaScript 生态系统已经有了多种 Promise 实现。尽管它们在内部的表示方式不同，但至少所有类 Promise 的对象都实现了 _Thenable_ 接口。thenable 对象实现了 [`.then()`](/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/Promise/then) 方法，该方法被调用时需要传入两个回调函数，一个用于 Promise 被兑现时调用，一个用于 Promise 被拒绝时调用。Promise 也是 thenable 对象。

为了与现有的 Promise 实现进行交互，JavaScript 语言允许在 Promise 的位置使用 thenable 对象。例如，[`Promise.resolve`](/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/Promise/resolve) 方法不仅可以解析 Promise 对象，还可以追踪 thenable 对象。

```js
const aThenable = {
  then(onFulfilled, onRejected) {
    onFulfilled({
      // thenable 对象被兑现为另一个 thenable 对象
      then(onFulfilled, onRejected) {
        onFulfilled(42);
      },
    });
  },
};

Promise.resolve(aThenable); // 一个兑现值为 42 的 Promise
```

### Promise 并发

`Promise` 类提供了四个静态方法来促进异步任务的[并发](https://zh.wikipedia.org/wiki/并发计算)：

- {{jsxref("Promise.all()")}}

>   - : `Promise.all` 方法返回一个 Promise，当**所有**传入的 Promise 都被兑现时，该 Promise 也成功兑现；当**任意一个** Promise 被拒绝时，该 Promise 也被拒绝。

Reviewer: ```suggestion
  - : 在**所有**传入的 Promise 都被兑现时兑现；在**任意一个** Promise 被拒绝时拒绝。
```

- {{jsxref("Promise.allSettled()")}}

>   - : 当**所有**的 Promise 都被敲定时（不论是被兑现还是被拒绝），`Promise.allSettled` 方法返回一个已兑现的 Promise。

Reviewer: ```suggestion
  - : 在**所有**的 Promise 都被敲定时兑现。
```

- {{jsxref("Promise.any()")}}

>   - : 当**任意一个** Promise 被兑现时，`Promise.any` 方法返回一个成功兑现的 Promise；当**所有**的 Promise 都被拒绝时，`Promise.any` 方法返回一个被拒绝的 Promise。

Reviewer: ```suggestion
  - : 在**任意一个** Promise 被兑现时兑现；仅在**所有**的 Promise 都被拒绝时才会拒绝。
```

- {{jsxref("Promise.race()")}}

>   - : 当**任意一个** Promise 被敲定时（无论是被兑现还是被拒绝），`Promise.race` 方法返回一个已兑现的 Promise。换句话说，当**任意一个** Promise 被兑现时，返回一个成功兑现的 Promise；当**任意一个**的 Promise 被拒绝时，返回一个被拒绝的 Promise。

Reviewer: ```suggestion
  - : 在**任意一个** Promise 被敲定时敲定。换句话说，在**任意一个** Promise 被兑现时兑现；在**任意一个**的 Promise 被拒绝时拒绝。
```

所有这些方法都接受一个 Promise（确切地说是 [thenable](#thenable)）的[可迭代对象](/zh-CN/docs/Web/JavaScript/Reference/Iteration_protocols#可迭代协议)，并返回一个新的 Promise。它们都支持子类化，这意味着它们可以在 `Promise` 的子类上调用，结果将是一个属于子类类型的 Promise。为此，子类的构造函数必须实现与 [`Promise()`](/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/Promise/Promise) 构造函数相同的签名——接受一个以 `resolve` 和 `reject` 回调函数作为参数的单个 `executor` 函数。子类还必须有一个 `resolve` 静态方法，可以像 {{jsxref("Promise.resolve()")}} 一样调用，以将值解析为 Promise。

> 请注意，JavaScript 的本质上是[单线程的](/zh-CN/docs/Glossary/Thread)，因此在任何时刻，只有一个任务会被执行，尽管控制权可以在不同的 Promise 之间切换，从而使 Promise 的执行看起来是并发的。在 JavaScript 中，[并行执行](https://zh.wikipedia.org/wiki/并行计算)只能通过[ worker 线程](/zh-CN/docs/Web/API/Web_Workers_API)实现。

Reviewer: ```suggestion
请注意，JavaScript 的本质上是[单线程的](/zh-CN/docs/Glossary/Thread)，因此在任何时刻，只有一个任务会被执行，尽管控制权可以在不同的 Promise 之间切换，从而使 Promise 的执行看起来是并发的。在 JavaScript 中，[并行执行](https://zh.wikipedia.org/wiki/并行计算)只能通过 [worker 线程](/zh-CN/docs/Web/API/Web_Workers_API)实现。
```

## 构造函数

- {{jsxref("Promise/Promise", "Promise()")}}
  - : 创建一个新的 `Promise` 对象。该构造函数主要用于封装还没有添加 promise 支持的函数。

## 静态属性

- {{jsxref("Promise/@@species", "Promise[@@species]")}}
  - : 返回用于构造从 Promise 方法返回值的构造函数。

> ## 静态方法

Reviewer: Still need updates on this section, please refer mdn/content#27560, which have made some more changes :)

Author: OK, I will mark this PR as a draft until everything was done.

一个已经处于“已敲定”状态的 promise 也可以接收操作。在那种情况下，（如果没有问题的话）这个操作会被作为第一个异步操作被执行。注意，所有的 promise 都一定是异步的。因此，一个已经处于“已敲定”状态的 promise 中的操作只有 promise 链式调用的栈被清空且一个时间片段过去之后才会被执行。这种效果跟 `setTimeout(action, 10)` 特别相似。

```js
const promiseA = new Promise((resolve, reject) => {
  resolve(777);
});
// 此时，“promiseA” 已经敲定了
promiseA.then((val) => console.log("异步日志记录有值：", val));
console.log("立即记录");

// 按以下顺序产生输出：
// 立即记录
// 异步日志记录有值：777
```

> ### Thenables

Reviewer: ```suggestion
### Thenable
```

> 在 Promise 成为 JavaScript 语言的一部分之前，JavaScript 生态系统已经实现了多个 Promise 实现。尽管它们在内部的表示方式不同，但至少所有类 Promise 的对象都实现了 _Thenable_ 接口。一个 thenable 对象实现了 [`.then()`](/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/Promise/then) 方法，该方法被调用时需要传入两个回调函数，一个用于 Promise 被兑现时调用，一个用于 Promise 被拒绝时调用。Promise 也是 thenable 对象。

Reviewer: ```suggestion
在 Promise 成为 JavaScript 语言的一部分之前，JavaScript 生态系统已经有了多种 Promise 实现。尽管它们在内部的表示方式不同，但至少所有类 Promise 的对象都实现了 _Thenable_ 接口。thenable 对象实现了 [`.then()`](/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/Promise/then) 方法，该方法被调用时需要传入两个回调函数，一个用于 Promise 被兑现时调用，一个用于 Promise 被拒绝时调用。Promise 也是 thenable 对象。
```

为了与现有的 Promise 实现进行交互，JavaScript 语言允许在 Promise 的位置使用 thenable 对象。例如，[`Promise.resolve`](/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/Promise/resolve) 方法不仅可以解析 Promise 对象，还可以追踪 thenable 对象。

```js
const aThenable = {
  then(onFulfilled, onRejected) {
    onFulfilled({
      // thenable 对象被兑现为另一个 thenable 对象
      then(onFulfilled, onRejected) {
        onFulfilled(42);
      },
    });
  },
};

Promise.resolve(aThenable); // 一个兑现值为 42 的 Promise
```

### Promise 并发

`Promise` 类提供了四个静态方法来促进异步任务的[并发](https://zh.wikipedia.org/wiki/并发计算)：

- {{jsxref("Promise.all()")}}

>   - : 当**所有**的 Promise 都被兑现时，`Promise.all` 方法返回一个成功兑现的 Promise；当**任意一个** Promise 被拒绝时，`Promise.all` 方法返回一个被拒绝的 Promise。

Reviewer: The description is not correct. The method would return a promise _even if_ all the passed promises is not settled.

>   - : 当**所有**的 Promise 都被兑现时，`Promise.all` 方法返回一个成功兑现的 Promise；当**任意一个** Promise 被拒绝时，`Promise.all` 方法返回一个被拒绝的 Promise。

Author: ```suggestion
  - : `Promise.all` 方法返回一个 Promise，当**所有**传入的 Promise 都被兑现时，该 Promise 也成功兑现；当**任意一个** Promise 被拒绝时，该 Promise 也被拒绝。
```

- {{jsxref("Promise.allSettled()")}}
  - : 当**所有**的 Promise 都被敲定时（不论是被兑现还是被拒绝），`Promise.allSettled` 方法返回一个已兑现的 Promise。

> - {{jsxref("Promise.any()")}}
>   - : 当**任意一个** Promise 被兑现时，`Promise.race` 方法返回一个成功兑现的 Promise；当**所有**的 Promise 都被拒绝时，`Promise.race` 方法返回一个被拒绝的 Promise。
> - {{jsxref("Promise.race()")}}
>   - : 当**任意一个** Promise 被敲定时（无论是被兑现还是被拒绝），`Promise.any` 方法返回一个已兑现的 Promise。换句话说，当**任意一个** Promise 被兑现时，返回一个成功兑现的 Promise；当**所有**的 Promise 都被拒绝时，返回一个被拒绝的 Promise。

Author: ```suggestion
- {{jsxref("Promise.any()")}}
  - : 当**任意一个** Promise 被兑现时，`Promise.any` 方法返回一个成功兑现的 Promise；当**所有**的 Promise 都被拒绝时，`Promise.any` 方法返回一个被拒绝的 Promise。
- {{jsxref("Promise.race()")}}
  - : 当**任意一个** Promise 被敲定时（无论是被兑现还是被拒绝），`Promise.race` 方法返回一个已兑现的 Promise。换句话说，当**任意一个** Promise 被兑现时，返回一个成功兑现的 Promise；当**任意一个**的 Promise 被拒绝时，返回一个被拒绝的 Promise。
```

> 所有这些方法都接受一个 Promise（确切地说是 [thenable](#thenables)）的[可迭代对象](/zh-CN/docs/Web/JavaScript/Reference/Iteration_protocols#可迭代协议)，并返回一个新的 Promise。它们都支持子类化，这意味着它们可以在 `Promise` 的子类上调用，结果将是一个属于子类类型的 Promise。为此，子类的构造函数必须实现与 [`Promise()`](/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/Promise/Promise) 构造函数相同的签名——接受一个以 `resolve` 和 `reject` 回调函数作为参数的单个 `executor` 函数。子类还必须有一个 `resolve` 静态方法，可以像 {{jsxref("Promise.resolve()")}} 一样调用，以将值解析为 Promise。

Reviewer: ```suggestion
所有这些方法都接受一个 Promise（确切地说是 [thenable](#thenable)）的[可迭代对象](/zh-CN/docs/Web/JavaScript/Reference/Iteration_protocols#可迭代协议)，并返回一个新的 Promise。它们都支持子类化，这意味着它们可以在 `Promise` 的子类上调用，结果将是一个属于子类类型的 Promise。为此，子类的构造函数必须实现与 [`Promise()`](/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/Promise/Promise) 构造函数相同的签名——接受一个以 `resolve` 和 `reject` 回调函数作为参数的单个 `executor` 函数。子类还必须有一个 `resolve` 静态方法，可以像 {{jsxref("Promise.resolve()")}} 一样调用，以将值解析为 Promise。
```

> 请注意，JavaScript 的本质上是[单线程的](/zh-CN/docs/Glossary/Thread)，因此在任何时刻，只有一个任务会被执行，尽管控制权可以在不同的 Promise 之间切换，从而使 Promise 的执行看起来是并发的。在 JavaScript 中，[并行执行](https://zh.wikipedia.org/wiki/并行计算)只能通过[工作线程](/zh-CN/docs/Web/API/Web_Workers_API)实现。

Reviewer: Is there any reference about the translation of "Worker"

Author: Nope. According to the link, "worker 线程" may be more proper

> 请注意，JavaScript 的本质上是[单线程的](/zh-CN/docs/Glossary/Thread)，因此在任何时刻，只有一个任务会被执行，尽管控制权可以在不同的 Promise 之间切换，从而使 Promise 的执行看起来是并发的。在 JavaScript 中，[并行执行](https://zh.wikipedia.org/wiki/并行计算)只能通过[工作线程](/zh-CN/docs/Web/API/Web_Workers_API)实现。

Author: ```suggestion
请注意，JavaScript 的本质上是[单线程的](/zh-CN/docs/Glossary/Thread)，因此在任何时刻，只有一个任务会被执行，尽管控制权可以在不同的 Promise 之间切换，从而使 Promise 的执行看起来是并发的。在 JavaScript 中，[并行执行](https://zh.wikipedia.org/wiki/并行计算)只能通过[ worker 线程](/zh-CN/docs/Web/API/Web_Workers_API)实现。
```

## 构造函数

- {{jsxref("Promise/Promise", "Promise()")}}
  - : 创建一个新的 `Promise` 对象。该构造函数主要用于封装还没有添加 promise 支持的函数。

## 静态属性

- {{jsxref("Promise/@@species", "Promise[@@species]")}}
  - : 返回用于构造从 Promise 方法返回值的构造函数。

## 静态方法

- {{jsxref("Promise.all()")}}

  - : 等待所有的 Promise 被兑现，或者有任何一个被拒绝。

    如果返回的 Promise 被兑现，那么它将以一个聚合数组的形式返回已兑现的 Promise 的值，顺序与包含多个 Promise 的可迭代对象中定义的顺序相同。

    如果有任何一个 Promise 被拒绝，那么返回的 Promise 就会被拒绝，并且拒绝的原因是可迭代对象中第一个被拒绝的 Promise 的原因。

- {{jsxref("Promise.allSettled()")}}

  - : 等待所有 Promise 都已敲定（可能是成功兑现或被拒绝）。

    返回一个 Promise，该 Promise 在所有给定的 Promise 都已成功兑现或被拒绝后被兑现，并且返回一个描述每个 Promise 结果的对象数组。

- {{jsxref("Promise.any()")}}

>   - : 接受一个 Promise 对象的可迭代对象，一旦可迭代对象中的任意一个 Promise 对象成功兑现，就会返回一个新的 Promise 对象，该新的 Promise 对象兑现该 Promise的 兑现值。

Reviewer: Actually, I don't like the desciption in en-US documation. That is, this method is not a "block method", it should return a promise immediately other than waiting for some promise is settled.

```js
console.log("start time: " + new Date().getTime())
promise = Promise.any([
	new Promise((resolve, reject) => {
		setTimeout(() => {
			resolve('foo')
		}, 3000)
	})
])

console.log("returned time: " + new Date().getTime())

promise.then((value) => {
	console.log("resolved time: " + new Date().getTime())
	console.log(value)
	// expected output: "foo"
})
```

![image](https://github.com/mdn/translated-content/assets/15844309/bb3fd274-e366-43ec-ba93-e7a7589d1562)

```suggestion
  - : 接受一个 Promise 对象的可迭代对象，一旦可迭代对象中的任意一个 Promise 对象成功兑现，就会返回一个新的 Promise 对象，该新的 Promise 对象兑现该 Promise 的兑现值。
```

Author: > Actually, I don't like the desciption in en-US documation. That is, this method is not a "block method", it should return a promise immediately other than waiting for some promise is settled.

I totally agree with you. I do realize that the English version has many problems, as I said in the initiation note of this PR. Considering that there is not much content in this chapter, and I need to compare all the contents of this chapter to have a deeper understanding of Promise, the modification of the original text and the update of the translation will be synchronized in the first round of all documents starting again.

Reviewer: filed mdn/content#27515

Reviewer: The issue has been resolved, could you update the content here @JasonLamv-t?

Author: > The issue has been resolved, could you update the content here @JasonLamv-t?

Sure.

- {{jsxref("Promise.race()")}}

> -

Reviewer: ```suggestion
```

    }
  }, 500);
}

function determineParity(value) {
  const isOdd = value % 2 === 1;
  return { value, isOdd };
}

function troubleWithGetNumber(reason) {
  const err = new Error("获取数据时遇到问题", { cause: reason });
  console.error(err);
  throw err;
}

function promiseGetWord(parityInfo) {
  return new Promise((resolve, reject) => {
    const { value, isOdd } = parityInfo;
    if (value >= THRESHOLD_A - 1) {
      reject(`还是太大了：${value}`);
    } else {
      parityInfo.wordEvenOdd = isOdd ? "奇数" : "偶数";
      resolve(parityInfo);
    }
  });
}

new Promise(tetheredGetNumber)
  .then(determineParity, troubleWithGetNumber)
  .then(promiseGetWord)
  .then((info) => {
    console.log(`得到了：${info.value}, ${info.wordEvenOdd}`);
    return info;
  })
  .catch((reason) => {
    if (reason.cause) {
      console.error("已经在前面处理过错误了");
    } else {
      console.error(`运行 promiseGetWord() 时遇到问题: ${reason}`);
    }
  })
  .finally((info) => console.log("所有回调都完成了"));
```

### 高级示例

本例展示了 `Promise` 的一些机制。`testPromise()` 方法在每次点击 {{HTMLElement("button")}} 按钮时被调用，该方法会创建一个 promise 对象，使用 {{domxref("setTimeout()")}} 让 `Promise` 等待 1-3 秒不等的时间来兑现计数结果（从 1 开始的数字）。使用 `Promise` 构造函数来创建 promise。

> 通过使用 {{jsxref("Promise.prototype.then()","p1.then()")}} 设置兑现回调函数，并在其中记录 Promise 的兑现，这些日志显示了方法的同步部份是如何与 Promise 的异步完成是如何解耦的。

Author: ```suggestion
通过使用 {{jsxref("Promise.prototype.then()","p1.then()")}} 设置兑现回调函数，并在其中记录 Promise 的兑现，这些日志显示了方法的同步代码是如何与 Promise 的异步完成是如何解耦的。
```

> 示例代码可以在 NodeJS 下运行。请通过查看实际发生的错误来理解代码。若要提高错误发生的概率，请该改变 `threshold` 的值。

Reviewer: Please update the line 253.

  setTimeout(() => {
    const randomInt = Date.now();
    const value = randomInt % 10;
    if (value < THRESHOLD_A) {
      resolve(value);
    } else {
      reject(`太大了: ${value}`);
    }
  }, 500);
}

function determineParity(value) {
  const isOdd = value % 2 === 1;
  return { value, isOdd };
}

function troubleWithGetNumber(reason) {
  const err = new Error("获取数据时遇到问题", { cause: reason });
  console.error(err);
  throw err;
}

function promiseGetWord(parityInfo) {
  return new Promise((resolve, reject) => {
    const { value, isOdd } = parityInfo;
    if (value >= THRESHOLD_A - 1) {
      reject(`还是太大了: ${value}`);
    } else {
      parityInfo.wordEvenOdd = isOdd ? "奇数" : "偶数";
      resolve(parityInfo);
    }
  });
}

new Promise(tetheredGetNumber)
  .then(determineParity, troubleWithGetNumber)
  .then(promiseGetWord)
  .then((info) => {
    console.log(`得到了：${info.value}, ${info.wordEvenOdd}`);
    return info;
  })
  .catch((reason) => {
    if (reason.cause) {
      console.error("已经在前面处理过错误了");
    } else {
      console.error(`运行 promiseGetWord() 时遇到问题: ${reason}`);
    }
  })
  .finally((info) => console.log("所有回调都完成了"));
```

> ### 高级示例

Reviewer: please update the line 305.

```js
"use strict";

let promiseCount = 0;

function testPromise() {
  const thisPromiseCount = ++promiseCount;
  const log = document.getElementById("log");
  // 开始
  log.insertAdjacentHTML("beforeend", `${thisPromiseCount}) Started<br>`);
  // 我们创建一个新的 Promise：我们承诺在等待 3 秒后，兑现从 1 开始计数的数字
  const p1 = new Promise((resolve, reject) => {
    // 执行器函数被调用，并具有解决或拒绝该 Promise 的能力
    log.insertAdjacentHTML(
      "beforeend",
      `${thisPromiseCount}) Promise constructor<br>`,
    );
    // 这只是一个创建异步操作的示例
    setTimeout(() => {
      // We fulfill the promise
      resolve(thisPromiseCount);
    }, Math.random() * 2000 + 1000);
  });

  // 我们使用 then() 来定义 Promise 被解决时的操作，
  // 并使用 catch() 来定义 Promise 被拒绝时的操作
  p1.then((val) => {
    // 打印兑现值
    log.insertAdjacentHTML("beforeend", `${val}) Promise fulfilled<br>`);
  }).catch((reason) => {
    // 打印拒绝原因
    console.log(`Handle rejected promise (${reason}) here.`);
  });
  // 结束
  log.insertAdjacentHTML("beforeend", `${thisPromiseCount}) Promise made<br>`);
}

const btn = document.getElementById("make-promise");
btn.addEventListener("click", testPromise);
```

#### 结果

> {{EmbedLiveSample("Advanced_Example", "500", "200")}}

Reviewer: ```suggestion
{{EmbedLiveSample("高级示例", "500", "200")}}
```
